aptos-move/framework/src/natives/signature.rs

```diff
@@ -18,7 +19,123 @@ use move_deps::{
+            Some(Ok(v)) => v,
+        };
+
+        // Pop each Value from the popped Vec<Value>, cast it as a Vec<u8>, and push it to a Vec<Vec<u8>>
+        let mut vec_vec = vec![];
+        for value in value_vec {
+            let vec = match value.value_as::<$t>() {
+                Err(e) => return Err(e),
+                Ok(v) => v,
+            };
+            vec_vec.push(vec);
+        }
+
+        vec_vec
+    }};
+}
+
+pub fn native_bls12381_aggregate_unvalidated_pubkeys(
+    _context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(_ty_args.is_empty());
+    debug_assert!(arguments.len() == 1);
+
+    // TODO: replace with proper gas cost
+    let cost = GasCost::new(super::cost::APTOS_LIB_TYPE_OF, 1).total();
+
+    // Parses a Vec<Vec<u8>> of all serialized public keys
+    let pks_serialized = pop_vec_arg!(arguments, Vec<u8>);
+    let mut pks = vec![];
+
+    for pk_bytes in pks_serialized {
+        // NOTE(Gas): O(|pk_bytes|) deserialization cost
+        let pk = match bls12381::PublicKey::try_from(&pk_bytes[..]) {
+            Ok(key) => key,
+            // If PK does not deserialize correctly, return None.
+            Err(_) => return Ok(NativeResult::ok(cost, smallvec![none_option()])),
+        };
+
+        pks.push(pk);
+    }
+
+    // If zero PKs were given as input, return None.
+    if pks.is_empty() {
+        return Ok(NativeResult::ok(cost, smallvec![none_option()]));
+    }
+
+    // Aggregate the public keys (this will NOT group-check the individual PKs)
+    let aggpk =
+        // NOTE(Gas): O(|pks|) cost: |pks| elliptic curve additions
+        match bls12381::PublicKey::aggregate(pks.iter().collect::<Vec<&bls12381::PublicKey>>()) {
+            Ok(aggpk) => aggpk,
+            Err(_) => return Ok(NativeResult::ok(cost, smallvec![none_option()])),
+        };
+
+    Ok(NativeResult::ok(
+        cost,
+        smallvec![some_option(aggpk.to_bytes().to_vec())],
+    ))
+}
+
+pub fn native_bls12381_validate_pubkey(
+    _context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    debug_assert!(_ty_args.is_empty());
+    debug_assert!(arguments.len() == 1);
+
+    // TODO: replace with proper gas cost
+    let cost = GasCost::new(super::cost::APTOS_LIB_TYPE_OF, 1).total();
+
+    let pk_bytes = pop_arg!(arguments, Vec<u8>);
+
+    // NOTE(Gas): O(|pk_bytes|) deserialization cost
+    let public_key = match bls12381::PublicKey::try_from(&pk_bytes[..]) {
+        Ok(key) => key,
+        Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
+    };
+
+    // NOTE(Gas): O(1) cost: uses endomorphisms for performing faster subgroup checks
+    let valid = public_key.group_check().is_ok();
+
+    Ok(NativeResult::ok(cost, smallvec![Value::bool(valid)]))
+}
+
+pub fn native_bls12381_verify_proof_of_possession(
```

Comment:
it's not only verify the PoP but also verify the public key can be deserialized?

Reply:
Yes: A successfully-verified PoP includes deserializing the PK successfully and checking it is a group element in the prime-order subgroup.

Reply:
I updated the docs to better reflect this.